src/persist/src/operators/persistent_upsert.rs

```diff
@@ -132,7 +222,60 @@ where
             },
         );
 
-        result_stream
+        let (new_upsert_oks, _new_upsert_persist_errs) =
+            new_upsert_oks.persist("upsert_state", persist_config.write_handle);
+
+        new_upsert_oks.concat(&restored_upsert_oks)
+    }
+}
+
+/// Ingests differential updates, consolidates them, and emits a final `HashMap` that contains the
+/// consolidated upsert state.
+struct DifferentialStateIngester<K, V> {
+    differential_state: HashMap<(K, V), isize>,
+}
+
+impl<K, V> DifferentialStateIngester<K, V>
+where
+    K: Hash + Eq + Clone + Debug,
+    V: Hash + Eq + Debug,
+{
+    fn new() -> Self {
+        DifferentialStateIngester {
+            differential_state: HashMap::new(),
+        }
+    }
+
+    fn add_update(&mut self, update: ((K, V), u64, isize)) {
+        let ((k, v), _ts, diff) = update;
+
+        self.differential_state
+            .entry((k, v))
+            .and_modify(|existing_diff| *existing_diff += diff)
+            .or_insert(diff);
+    }
+
+    fn finish(mut self) -> HashMap<K, V> {
+        self.differential_state.retain(|_k, diff| *diff > 0);
+
+        let mut state = HashMap::new();
```

Comment:
I think we could, though we would have the problem that a specific operator needs to figure out which keys it is responsible for. And then maybe still ingest data into a HashMap so that it can be used. Unless we want the operator to be wholly backed by a `TraceReader`. Note: a previous upsert operator did use a differential arrangement as its backing store but was replaced by the current HashMap-based operator because of performance and memory-usage reasons.